src/components/Item.tsx

```diff
@@ -1,45 +1,70 @@
 import React from "react";
-import { View, Image, Text, ImageProps, StyleSheet } from "react-native";
+import {
+  View,
+  Image,
+  Text,
+  StyleSheet,
+  type ImageURISource,
+} from "react-native";
 import Octicons from "@expo/vector-icons/Octicons";
 import { Typography } from "@/theme/typography";
 import { ColorTheme } from "@/theme/colors";
 
 export const Item = (props: ItemProps) => {
-  const { heading, subtitle, body, headingSize, ...avatarProps } = props;
-  const bodyProps = { heading, subtitle, body, headingSize };
+  const { heading, subtitle, body, headingSize, badgeName, iconName, source } =
+    props;
+
   return (
     <View style={ItemStyles.row}>
-      <ItemAvatar {...avatarProps} />
-      <ItemBody {...bodyProps} />
+      <ItemAvatar badgeName={badgeName} iconName={iconName} source={source} />
+      <ItemBody
+        heading={heading}
+        subtitle={subtitle}
+        body={body}
+        headingSize={headingSize}
+      />
```

Comment:
`ItemProps` is currently defined like so: 
`export type ItemProps = ItemAvatarProps & ItemBodyProps;`

With this proposed destructuring of the props, the caller _might_ have specified a property in `ItemAvatarProps` or `ItemBodyProps` that does not get destructured and assigned here. This seems like a tight-coupling between those types and this function. 

Ex: If I added a prop to `ItemAvatarProps`, I'd need to update this function to destructure that new prop in this function, and assign it to the `ItemAvatar` component. Whereas previously, it would automatically propagate down without needing to modify this function.

This is probably my inexperience with Typescript/Javascript, but does introducing this destructuring have benefits beyond what's immediately visible here? I don't see any type-errors on this particular function at the moment, so I'm mostly curious!

Reply:
> With this proposed destructuring of the props, the caller might have specified a property in ItemAvatarProps or ItemBodyProps 

BodyProps is created on line 9 and then blown up several lines later (side note: this introduces a runtime cost for no gain). So BodyProps isn't actually giving us any flexibility it's just a data structure that got created and destroyed. On the other hand `avatarProps` actually uses the parameter (...rest) operator which is intended to capture many params or unknown params. Which illustrates my next point: if it's easy to miss how strange this exception is even during our code review of one component, imagine what doing this everywhere throughout a piece of software does. Furthermore, how would the person importing `<Item />` even know that when they passed misc. params they were going to be spread to avatar rather than `body`? Why even either of them, why wouldn't they be passed to the container of `<Item />` which is what I would expect when importing a component and then passing it a style. See what I'm getting at here? We lose so much explicitness and predictability that it becomes difficult to understand what our interfaces are and why. Exceptions are everywhere and they are arbitrary. In the long run, take the following pseudocode and imagine having to dig through multiple files to find what's going in our application and still struggling to understand:

```
PageWrapper((...params) ->
   PageContainer((...params) ->
     PageProvider((...params) ->
        Heading(({title} = params), ...otherParams) ->
        Body(({body} = params, ...otherParams)) ->
```

A piece of software that has had this operator seeded everywhere to pass parameters becomes difficult to refactor without introducing bugs. Newcomers and OSS contributors think it's okay, so they'll use it, adding to both runtime cost and implicitness. Development starts to grind to a halt. Newcomers eventually find it more difficult to get onboarded. Tech leads will huddle on it and come to the conclusion that they shouldn't have built an app that way and your next big tech debt cleanup of the quarter will be (extremely painfully) removing doing this where it was never absolutely necessary. There's a significant runtime cost in doing this everywhere. It makes using TypeScript much more challenging. This is aside from the numerous other issues it introduces including the most challenging and hidden bugs that I've ever had to remove in a Javascript application.

Now here's ChatGPT on it because I'm lazy:

> Lack of Explicitness: One of the most substantial arguments against using the spread operator for props is that it makes the code less explicit. Readers can't easily tell what props are being passed to a component without tracing back to the source of the object being spread. This can make code maintenance and understanding more difficult.
> 
> Unintended Props: There's a risk of passing extra props unintentionally if the object contains more properties than you intend to pass to the component. This can lead to unexpected behavior and bugs that are hard to trace because the component might receive props it shouldn't have or wasn't designed to handle.
> 
> Overwrites and Duplication: If you use the spread operator and then specify another prop afterward, the explicitly defined prop will overwrite any prop from the spread with the same name. This behavior can introduce subtle bugs if a developer isn't aware of the properties in the spread object.
> 
> <Component {...props} someProp="value" />
> In the example above, if props has a someProp property, it will be overwritten by "value", which might not be the intended behavior.
> 
> Performance Overhead: While modern JavaScript engines are optimized and efficient, constantly creating new objects and spreading them can introduce some performance overhead, especially if done within frequently rendered components.
> 
> Type Safety Concerns: If you're using TypeScript or Flow, spreading props can make it more challenging to maintain strict type safety. It becomes harder to ensure that the component receives props of the right type or that required props are indeed provided.
> 
> Security Concerns: In some scenarios, spreading props might lead to potential security risks. For instance, if you're spreading an object that contains data fetched from an API directly into a component, there might be a risk of injecting unintended or harmful props.
> 
> Component's Contract: Every component has an implicit or explicit "contract" about what props it expects and how it uses them. Using the spread operator can obscure this contract, making the component's behavior less predictable.
> 
> Difficulty in Refactoring: When components are tightly coupled with the structure of their props, it makes refactoring more difficult. If you decide to change the shape of the object or rename properties, it can break components where the spread operator was used, leading to more extensive code changes.
> 
> In conclusion, while the spread operator provides convenience and can make code more concise, it comes with trade-offs that can impact code quality, maintainability, and even performance in a React app. It's essential to use it judiciously and be aware of its potential pitfalls.

Reply:
Fascinating! Thanks for the thorough explanation here. 

React is a crazy beast that I'm still trying to understand fully. With my knowledge so far, this kinda makes sense, kinda doesn't :P 

I'm sure it'll make more sense as I'm getting more hands-on with the view rendering. Definitely trust you here, just wanted to start getting an understanding!